addons/point_of_sale/models/pos_payment_method.py

```diff
@@ -88,12 +114,92 @@ def _is_write_forbidden(self, fields):
         whitelisted_fields = {'sequence'}
         return bool(fields - whitelisted_fields and self.open_session_ids)
 
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals.get('payment_method_type', False):
+                self._force_payment_method_type_values(vals, vals['payment_method_type'])
+        return super().create(vals_list)
+
     def write(self, vals):
         if self._is_write_forbidden(set(vals.keys())):
             raise UserError(_('Please close and validate the following open PoS Sessions before modifying this payment method.\n'
                             'Open sessions: %s', (' '.join(self.open_session_ids.mapped('name')),)))
-        return super(PosPaymentMethod, self).write(vals)
+
+        if 'payment_method_type' in vals:
+            self._force_payment_method_type_values(vals, vals['payment_method_type'])
+            return super().write(vals)
+
+        pmt_terminal = self.filtered(lambda pm: pm.payment_method_type == 'terminal')
+        pmt_qr = self.filtered(lambda pm: pm.payment_method_type == 'qr_code')
+        not_pmt = self - pmt_terminal - pmt_qr
+
+        res = True
+        forced_vals = vals.copy()
+        if pmt_terminal:
+            self._force_payment_method_type_values(forced_vals, 'terminal', True)
+            res = super(PosPaymentMethod, pmt_terminal).write(forced_vals) and res
+        if pmt_qr:
+            self._force_payment_method_type_values(forced_vals, 'qr_code', True)
+            res = super(PosPaymentMethod, pmt_qr).write(forced_vals) and res
+        if not_pmt:
+            res = super(PosPaymentMethod, not_pmt).write(vals) and res
+
+        return res
+
+    @staticmethod
+    def _force_payment_method_type_values(vals, payment_method_type, if_present=False):
+        if payment_method_type == 'terminal':
+            disabled_fields_name = ['qr_code_method']
+        elif payment_method_type == 'qr_code':
+            disabled_fields_name = ['use_payment_terminal']
+        else:
+            disabled_fields_name = ['use_payment_terminal', 'qr_code_method']
+        if if_present:
+            for name in disabled_fields_name:
+                if name in vals:
+                    vals[name] = False
+        else:
+            for name in disabled_fields_name:
+                vals[name] = False
```

Comment:
May I know why this block of code needs to be this complicated?

Reply:
I took inspiration from this: https://github.com/odoo/odoo/blob/17.0/addons/pos_online_payment/models/pos_payment_method.py#L77
Which clears all the fields that are not used by online payment when it is selected to ensure there is no misconfiguration
In this PR I introduced a new selection to define if the payment method uses "no integration", a terminal or a QR.
So instead of a boolean in online payment I have multiple possible states and each of them have their own fields to clear when they are selected.